Improve Python FAQ

jcfr opened the pull request:
This pull request adds the following entries:
* What is the Slicer Python Environment?
* What is the PythonSlicer executable?
* What is the Python Console?
* How to access Slicers Python API via an external program while Slicer is running?

Additionally, it update the Slicer API to document the following:
* `slicer.app`
* `slicer.mrmlScene`
* `slicer.modules`
* `slicer.moduleNames`

Summary of this pull request also posted on discourse at https://discourse.slicer.org/t/updated-python-faq-slicer-api-documentation/29967

<details><summary>Click here for preview screenshots</summary>

| | |
|--|--|
| Slicer API | ![image](https://github.com/Slicer/Slicer/assets/219043/312a9b3e-7e9c-4dd9-a5ac-7c71f858587d) |
| Python FAQ | ![image](https://github.com/Slicer/Slicer/assets/219043/9909a772-1c5d-4ad3-a0b9-99654cd02fba) |
|   | ![image](https://github.com/Slicer/Slicer/assets/219043/30749ce1-615b-4771-977f-2a88e66a7b28) |


</details>

jamesobutler commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+
+`PythonSlicer` is an executable provided in the `bin` directory of the Slicer installation.
+It’s the same version of Python available through the Slicer application, but it acts like
+a normal `python` (or `python3`) executable in terms of running scripts and can be used t
```
```suggestion
a normal `python` (or `python3`) executable in terms of running scripts and can be used to
```

jamesobutler commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+the [user interface](/user_guide/user_interface.md).
+
+
+It allows to access to the running application context including the application instance,
```
```suggestion
It allows to access the running application context including the application instance,
```

lassoan replied:
Do you mean *Python* context?
```suggestion
It allows to access to the running Python context including the application instance,
```

jamesobutler commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+  requests per second (e.g., continuous data streaming) or for clients that only have access to
+  sockets and don't want to reimplement complex protocols like http. In most cases, it performs well
+  for sending requests in the range of 10 to 100 requests per second.
+  There is a native Python implementation: [pyigtl](https://pypi.org/project/pyigtl/).
```
Can you describe if you would use `pyigtl` in addition to SlicerOpenIGTLink or as an alternative to SlicerOpenIGTLink? Any reason for picking one over the other?  Maybe provide info regarding if SlicerOpenIGTLink can be manipulated through python as well even though OpenIGTLinkIF is originally written in C++.

lassoan replied:
In Slicer SlicerOpenIGTLink should be used (pyigtl just uses Python threading, which is not really parallelization).
pyigtl can be used everywhere else.

jamesobutler replied:
Since this is Slicer documentation, does it make sense to include essentially general OpenIGTLink usage here? It seems like different uses of IGTLink whether with Slicer or some pure Python implementation might be better in the OpenIGTLink documentation.

lassoan replied:
I agree that conceptually it is redundant to write about these interprocess communication options in the Slicer core documentation, but it has huge practical value to see a summary of all options at one place (similarly how we write about extensions that provide registration modules or DICOM plugins). The descriptions are fairly high-level, so they should not require a lot of maintenance.

jamesobutler commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+* {attr}`slicer.moduleNames`
+
+:::{hint}
+Running scripts (or code) using the command-line option `--python-script` (or `pythoin-code`) is equivalent
```
```suggestion
Running scripts (or code) using the command-line option `--python-script` (or `python-code`) is equivalent
```

lassoan commented on Base/Python/slicer/__init__.py:
```diff
@@ -1,4 +1,108 @@
+
+  .. warning::
+
+    This attribute is set only in the context of a running Slicer application.
```
It is not very clear what we mean by "in the context of a running Slicer application". Something like this may be better:
```suggestion
    This attribute is only set in the Python environment embedded in the Slicer main application SlicerApp-real (launched by Slicer executable). The attribute is not set in the Python environment of python-real (launched by PythonSlicer executable).
```

lassoan commented on Base/Python/slicer/__init__.py:
```diff
@@ -1,4 +1,108 @@
+
+  .. warning::
+
+    This attribute is set only in the context of a running Slicer application.
```
Instead of repeating this somewhat vague warninng several times, it could be better to create a subsection for all these attributes that are set only in the main Slicer GUI application.

jcfr replied:
Ditto. Here is an issue tracking this:
* https://github.com/Slicer/Slicer/issues/7016

I the mean time, I will update the consistently update the waning based on your suggestion.

lassoan commented on Base/Python/slicer/__init__.py:
```diff
@@ -1,4 +1,108 @@
+      >>> isinstance(slicer.modules.VectorToScalarVolumeInstance, slicer.ScriptedLoadableModule.ScriptedLoadableModule)
+      True
+
+  * the attribute ``<moduleName>Widget`` is set to the corresponding instance
```
What if the Python module widget is reloaded? Does it point to the current widget instance or the first? The first module widget may become obsolete if the module widget is reloaded.

lassoan commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+## What is the PythonSlicer executable?
+
+`PythonSlicer` is an executable provided in the `bin` directory of the Slicer installation.
+It’s the same version of Python available through the Slicer application, but it acts like
```
```suggestion
It is the same version of Python available through the Slicer application, but it acts like
```

lassoan commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+
+`PythonSlicer` is an executable provided in the `bin` directory of the Slicer installation.
+It’s the same version of Python available through the Slicer application, but it acts like
+a normal `python` (or `python3`) executable in terms of running scripts and can be used t
```
PythonSlicer acts as a usual `python` executable in every sense, not just in term of running scripts. For example, it can be used as an interactive console, too. So, I'm not sure if we need to narrow its scope (we could just remove the "in term of..." part).

jcfr replied:
Very good point, I revisited the wording:

```
`PythonSlicer` is an executable provided in the `bin` directory of the Slicer installation.
It is a Python interpreter that allows access to all of the packages installed in Slicer.
This makes it possible to use the Slicer Python environment from outside the application, such
as for batch processing, command-line operations, terminal-based interactive session and
integration with IDEs.

This means you can use `PythonSlicer` as a replacement for the regular Python interpreter
(`python` or `python3`) to take advantage of the installed packages in the Slicer environment. 

:::{warning}
Please note that the running Python context is **not** available when using `PythonSlicer`.
As a result, many objects, including the application instance, the MRML scene and the loaded
modules, are not available.
:::
```

lassoan commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+ access all the packages installed in Slicer.
+
+This means you can use it in place of regular python to make use of the Slicer environment
+for batch processing and command-line operations.
```
Some examples would be great here.

jcfr replied:
Ditto. Here is an issue tracking this:
* https://github.com/Slicer/Slicer/issues/7015


Since depending on the needed capabilities batch processing may have to be implemented through the ``Python Console by leveraging `--python-script` and `--no-main-windows`, I expanded the `What is the Python Console` section with the following:


```
:::{hint}
Running scripts (or code) using the command-line option `--python-script` (or `python-code`)
is equivalent to running code in the Python Console.

Combined with the use of `--no-main-window`, this is useful for implementing batch processing
pipelines leveraging capabilities only available in the context of a running application.
For example, this applies to the [Segment Editor effects](/user_guide/modules/segmenteditor.html#effects).
:::
```

jcfr replied:
I think I found a nice compromise in the context of the API documentation:

![image](https://github.com/Slicer/Slicer/assets/219043/e16118e1-59d6-4e09-af8d-99294383a936)

lassoan commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+
+## What is the Python Console?
+
+This describes the [](/user_guide/user_interface.md#python-console) available through
```
Is the empty `[]` valid syntax?

jcfr replied:
It is indeed, it means that corresponding header is used in the generated text. 

> If no text is given or it is an auto-link, e.g. `[](#dest)` or `<project:#dest>`, then MyST will attempt to resolve an implicit text. For example, if the destination is a heading, then the heading text will be used as the link text, or if the destination is a figure/table then the caption will be used as the link text. Otherwise, the link text will be the destination itself.

See https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#explicit-vs-implicit-link-text

lassoan replied:
Thanks for the explanation.

lassoan reviewed (commented):
Thank you, these changes are very useful. I've added some inline comments.

pieper commented on Docs/developer_guide/python_faq.md:
```diff
@@ -2,6 +2,85 @@
+See [related discussion](https://discourse.slicer.org/t/rename-python-interactor-to-python-console/25897).
+:::
+
+## How to access Slicers Python API via an external program while Slicer is running?
```
Slicer's

pieper commented:
Thanks Jc!

I didn't read very carefully.  I suggest we merge, read it in context, and then make new PRs for any improvements.

jcfr commented:
Thanks everyone for the feedback :pray: 

To summarize, I think I addressed most of the concerns

I also created the following new issues to track future improvements:
* https://github.com/Slicer/Slicer/issues/7015
* https://github.com/Slicer/Slicer/issues/7016

If there are no other concerns, I would to move forward with integration.

> re: pyigtl

Should we improve the wording of the FAQ entry `## How to access Slicer's Python API via an external program while Slicer is running?` or are we happy with it ?

lassoan commented:
Just two small clarifications that could be done Beedie merging:

1. Can you describe what happens to the ...WidgetInstance attribute if a Python scripted loadables module if reloaded?
2. Add a comment that pyigtl is intended to be used outside Slicer, to send/receive data to/from Slicer.

jcfr commented:
> re: Add a comment that pyigtl is intended to be used outside Slicer, to stream data to/from Slicer.


```diff
diff --git a/Docs/developer_guide/python_faq.md b/Docs/developer_guide/python_faq.md
index 36a8081072..05cb02d553 100644
--- a/Docs/developer_guide/python_faq.md
+++ b/Docs/developer_guide/python_faq.md
@@ -69,9 +69,12 @@ There are several ways to access Slicer's Python API from an external program wh
 - [SlicerOpenIGTLink](https://github.com/openigtlink/SlicerOpenIGTLink): A lightweight socket-based
   protocol for real-time data transfer. This is useful for applications that need to send many
   requests per second (e.g., continuous data streaming) or for clients that only have access to
-  sockets and don't want to reimplement complex protocols like http. In most cases, it performs well
-  for sending requests in the range of 10 to 100 requests per second.
-  There is a native Python implementation: [pyigtl](https://pypi.org/project/pyigtl/).
+  sockets and prefer to avoid the complexity of protocols like HTTP. In most cases, it performs well
+  for sending requests at a rate of 10 to 100 requests per second.
+
+  A native Python implementation, [pyigtl](https://pypi.org/project/pyigtl/) , is available for
+  use outside the Slicer application. It can be used to both stream data from Slicer and stream
+  data to Slicer.
 
 - [SlicerJupyter](https://github.com/Slicer/SlicerJupyter#readme): A protocol for interacting with
   Slicer using standard Jupyter kernel protocol (simple protocol built on top of ZeroMQ middleware).
```

jcfr commented:
> re: Can you describe what happens to the ...WidgetInstance attribute if a Python scripted loadables module if reloaded?

I checked and is indeed updated.

![image](https://github.com/Slicer/Slicer/assets/219043/edae6eda-4c44-4da1-8dd0-80dc4f0cbf56)


https://github.com/Slicer/Slicer/blob/d4ba0f61075206109d3678d5d70c52bd741e64a3/Base/Python/slicer/util.py#L1364-L1367
